src/ds/flaglock.h

```diff
@@ -4,22 +4,40 @@
 
 #include <atomic>
 
+#ifndef NDEBUG
+#  include <thread>
+#endif
+
 namespace snmalloc
 {
   class FlagLock
   {
   private:
     std::atomic_flag& lock;
 
+#ifndef NDEBUG
+    std::thread::id owner{};
+#endif
   public:
     FlagLock(std::atomic_flag& lock) : lock(lock)
     {
       while (lock.test_and_set(std::memory_order_acquire))
+      {
+#ifndef NDEBUG
+        SNMALLOC_ASSERT(std::this_thread::get_id() != owner);
+#endif
```

Comment:
Does this need to be in `NDEBUG`, isn't `SNMALLOC_ASSERT` already guarded on `NDEBUG`?

Reply:
This test can never succeed.  Even if the lock is held by the current thread, the thing that you're looking at is a field that has, on entry to this function, been initialised to invalid and which cannot be concurrently modified.